tf2onnx/rewriter/rnn_utils.py

```diff
@@ -263,3 +267,189 @@ def get_weights_from_const_node(g, node):
         return None
 
     return RnnWeight(node, val, dtype)
+
+
+######################################################
+####      Utilities for bidirectional rnn      #######
+######################################################
+class ONNX_RNN_TYPE(Enum):
+    GRU = 0
+    LSTM = 1
+
+
+onnx_rnn_type_mapping = {
+    ONNX_RNN_TYPE.GRU: "GRU",
+    ONNX_RNN_TYPE.LSTM: "LSTM"
+}
+
+
+def find_bidirectional_rnns(g, ops, rnn_type):
+    """
+    Find possible bidirectional rnns, return: list of tuple,
+    Format of tuple is (fw onnx rnn node, bw onnx rnn node).
+    """
+    fw_rnns = defaultdict(list)
+    bw_rnns = defaultdict(list)
+    for n in g.get_nodes():
+        if n.type != onnx_rnn_type_mapping[rnn_type]:
+            continue
+
+        input_id = n.input[0]
+        temp = n.inputs[0]
+        is_bw = False
+        if temp.type == "Transpose":
+            input_id = temp.input[0]
+            temp = temp.inputs[0]
+
+        if utils.is_tf_reverse_op(temp):
+            input_id = temp.input[0]
+            is_bw = True
+
+        if is_bw:
+            # if output 0 is consumed and there is no reverse after the 1st output.
+            # it's not backward rnn.
+            if g.find_output_consumers(n.output[0]) and not get_reverse_nodes_after_y_output(g, n):
+                logger.warning("rnn %s following Reverse op isn't the part of bi-rnn.", n.name)
+                continue
+
+            logger.debug("find bw rnn %s", input_id)
+            bw_rnns[input_id].append(n)
+        else:
+            logger.debug("find fw rnn %s", input_id)
+            fw_rnns[input_id].append(n)
+
+    # fw_rnn and bw_rnn must share the same input
+    birnn_input = list(set(fw_rnns.keys()).intersection(bw_rnns.keys()))
+    bi_rnns = []
+    for inp in birnn_input:
+        fw_rnn = fw_rnns[inp]

```

Comment:
could we add a comment here: if fw and bw share same input and have same name scope, then they can be combine into one birnn? or modify the name "belong_to_birnn" into "is_same_rnn_name_scope"